# Please complete the task from Prompt #1 OR Prompt #2 below.

# **Prompt \# 1**

The goal of this exercise is to make a variant calling pipeline (python
or snakemake is preferred, but any general purpose language is
acceptable). Submit your workflow into a public git repository. Please
only do as much as you can in under three hours.

### **Minimum requirements:**

Workflow takes short read data (fastq files) as input and outputs
variant calls (a vcf file).

An example pipeline would be using bwa mem → samtools view/sort→
freebayes.

bwa: [[bwa]{.underline}](http://bio-bwa.sourceforge.net/bwa.shtml)

samtools: [[samtools]{.underline}](http://www.htslib.org/)

freebayes:
[[freebayes]{.underline}](https://github.com/freebayes/freebayes)

##### **Some example data you can use:**

Paired Illumina reads from a publicly available SARS-CoV-2 dataset ([SRA
accession
SRR15660643](https://www.ncbi.nlm.nih.gov/sra/?term=SRR15660643))
downsampled to 16000 paired reads (mapping and variant calling should
all be able to run in \<\<1 min). Gzipped fastq files for the paired
reads can be found as attachments in this
[[folder]{.underline}](https://drive.google.com/drive/folders/1jy1elsBSARjzpdf6_S3oVki58NwRGdO1)
(sample.R1.paired.fq.gz and sample.R2.paired.fq.gz).

The Wuhan-Hu-1 reference genome ([Genbank accession
MN908947.3](https://www.ncbi.nlm.nih.gov/nuccore/MN908947.3)). A fasta
file with this reference genome can be found as an attachment in this
[[folder]{.underline}](https://drive.google.com/drive/folders/1jy1elsBSARjzpdf6_S3oVki58NwRGdO1)
(MN908947.3.fasta).

### **Extra points:**

-   Use a real life data set of your choosing

-   Create unit tests

-   Automatically check and log potential errors

-   Create a Dockerfile that encapsulates all tools

-   Add QC, ex. MultiQC, FastQC.

-   Filter your VCF using QUAL or DP

-   Add Mark Duplicates

-   Annotate the VCF (ex. SNPeff or VEP)

-   Provide some interpretation of the vcf - e.g., thoughts on
    > contamination/mosaicism from heterozygosity ratios

# **Prompt \# 2**

The goal of this exercise is to produce a workflow that filters and
summarizes a [VCF
file](https://www.internationalgenome.org/wiki/Analysis/Variant%20Call%20Format/vcf-variant-call-format-version-40/)
(python or snakemake is preferred, but any general purpose language is
acceptable). Submit your workflow into a public git repository.

### **Minimum requirements:**

The workflow will accept a VCF file and perform the key operations
listed below on it. Please only do as much as you can in under three
hours. Feel free to use your own VCF or the one we've provided.

-   Filter by Quality \[provide justification for threshold\]

-   Filter by Depth \[provide justification for threshold\]

-   Filter by strand bias

-   Annotates the VCF (by VEP/SNPEff)

-   Provide summary stats for the VCF: \# of 0/1, 1/1, 0/0 variants,
    > het/homo ratio, \# of variants across annotated categories. Output
    > all the measurements you think will be useful.

-   Filter variants by genomic position range and compute average depth
    > of variants between them.

-   Plot heterozygosity vs depth and comment on filtration criteria used
    > and any observed patterns.

-   Plot heterozygosity vs Qscore and comment on filtration criteria
    > used and any observed patterns.

-   Get MAF from ExAC or GnoMAD (unless SnpEff already does that and if
    > it's a human sample)

##### **Some example data you can use:**

A sample VCF generated by synthetically mixing reads from two publicly
available SARS-CoV-2 datasets ([SRA accession
SRR15660643](https://www.ncbi.nlm.nih.gov/sra/?term=SRR15660643) and
[SRA accession
SRR15270309](https://www.ncbi.nlm.nih.gov/sra/?term=SRR15270309)) and
mapping/calling variants based on the reference genome below. This
sample vcf can be found as an attachment in this
[[folder]{.underline}](https://drive.google.com/drive/folders/1jy1elsBSARjzpdf6_S3oVki58NwRGdO1)
(sample.vcf).

The Wuhan-Hu-1 reference genome ([Genbank accession
MN908947.3](https://www.ncbi.nlm.nih.gov/nuccore/MN908947.3)). A fasta
file with this reference genome can be found as an attachment in this
[[folder]{.underline}](https://drive.google.com/drive/folders/1jy1elsBSARjzpdf6_S3oVki58NwRGdO1)
(MN908947.3.fasta).

### **Extra points:**

-   Use a real life data set of your choosing

-   Create unit tests

-   Automatically check and log potential errors

-   Create a Dockerfile that encapsulates all tools

-   Provide some interpretation of the vcf - e.g., thoughts on
    > contamination/mosaicism from heterozygosity ratios
